google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/startup/UnsupportedJavaVersionCheck.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.startup;
+
+import com.google.cloud.tools.intellij.appengine.cloud.AppEngineEnvironment;
+import com.google.cloud.tools.intellij.appengine.facet.AppEngineFacet;
+import com.google.cloud.tools.intellij.appengine.project.AppEngineAssetProvider;
+import com.google.cloud.tools.intellij.appengine.project.AppEngineProjectService;
+import com.google.cloud.tools.intellij.util.GctBundle;
+
+import com.intellij.notification.Notification;
+import com.intellij.notification.NotificationDisplayType;
+import com.intellij.notification.NotificationGroup;
+import com.intellij.notification.NotificationListener;
+import com.intellij.notification.NotificationType;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.module.EffectiveLanguageLevelUtil;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.LanguageLevelModuleExtension;
+import com.intellij.openapi.roots.ModifiableRootModel;
+import com.intellij.openapi.roots.ModuleRootManager;
+import com.intellij.openapi.startup.StartupActivity;
+import com.intellij.pom.java.LanguageLevel;
+import com.intellij.psi.xml.XmlFile;
+import com.intellij.psi.xml.XmlTag;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.swing.event.HyperlinkEvent;
+
+/**
+ * A StartupActivity that warns the user if they are using an unsupported java version.
+ */
+public class UnsupportedJavaVersionCheck implements StartupActivity {
+
+  private static final String RUNTIME_TAG = "runtime";
+  private static final String RUNTIME_TAG_JAVA_8 = "java8";
+  private static final String UPDATE_HREF = "#update";
+  private static final LanguageLevel HIGHEST_SUPPORTED_LANGUAGE_LEVEL = LanguageLevel.JDK_1_7;
+
+  @Override
+  public void runActivity(@NotNull Project project) {
+    List<Module> invalidModules = findAppEngineModulesUsingUnsupportedLanguageLevel(project);
+    if (!invalidModules.isEmpty()) {
+      warnUser(project, invalidModules);
+    }
+  }
+
+  private List<Module> findAppEngineModulesUsingUnsupportedLanguageLevel(Project project) {
+    Module[] projectModules = ModuleManager.getInstance(project).getModules();
+    List<Module> invalidModules = new ArrayList<>();
+
+    for (Module module : projectModules) {
+      // if it's not an app engine module, skip it
+      if (!hasAppEngineFacet(module)) {
+        continue;
+      }
+
+      @Nullable
```

Comment:
This is actually a valid use of the `@Nullable` anotation (it's declared as applicable to `ElementType.LocalVariable`).  It's a bit uncommon to use it in this way, but I think it's important to explicitly call out that since it's assigned to the result of a method that's annotated as nullable, it too is nullable. Without the annotation, I can too easily envision someone refactoring this method and not thinking to check that variable for a null value.